### DISCUSSION TITLE
how to improve performance

### DISCUSSION THREAD

**Discussion by jeffguorg** [libreswan/libreswan#1112] [Internals]
# System
OS: CentOS 7
Kernel: 3.10(stock) / 5.1.15(elrepo)
Libreswan Version: 3.10(stock)/3.25(built with [spec](https://git.centos.org/rpms/libreswan.git))
Hardware: Xeon 2620v4 + X520-SR2

# Target & Current status 
make full use of 10Gbps nic. 
But currently the bandwidth via ipsec is limited to 1.8Gbps when the app and ksoftirq is running 100% of CPU. 
And i just can't figure out how to reach ```5.25 Gbits/sec IPsec AES_GCM128 (esp=aes_gcm128-null)``` in the wiki

# Attempted methods
- Use simple config of one SA. (bandwidth is limited to 1.8 Gbps. )
- Start multiple SA and load balance on it. (works without problem). but we want a more simple solution)
- (current): trying to make nic-offload work.
   I see redhat has already back-ported the nic-offload code to RHEL7.5 and CentOS should also be patched. But the stock kernel(/boot/config-xxxxx) has no ESP_OFFLOAD options and ```ip xfrm state``` showed no ```offload dev xxx```option.
   So i installed the latest kernel from elrepo, which compiles offload part into kmod. and upgrade libreswan to 3.25. But still no luck, whether i modprobe esp4_offload or not. config used here is 
   ```
    conn myconn
        left=<leftip>
        right=<rightip>
        leftsubnet=10.0.0.0/16
        rightsubnet=32.18.108.11/24
        
        phase2alg=aes_gcm-null

        nic-offload=yes
        replay-window=128
        
        auth=secret
        auto=start
   ```

> update: pasted wrong config. it's correct now, using esp=aes_gcm-null

**Comment by letoams** [comment]
Do you have a nic card that supports nic offload?

**Comment by jeffguorg** [comment]
Sure. We checked the product brief and datasheets. Intel X520 10Gbps NICs should support nic offload and esp encapsulation. and we have a X520(82599ES). 
is there anyway i can check whether it support that in the system? i scammed across dmesg but didn't recognize something useful.

ref on Intel X520:
https://www.intel.com/content/www/us/en/embedded/products/networking/82599-10-gbe-controller-datasheet.html?asset=2377
https://www.kernel.org/doc/html/latest/networking/device_drivers/intel/ixgbe.html

> update: testing this post: https://blogs.oracle.com/linux/tips-and-tricks-for-ipsec-on-intel-10-gbe-nics-v2
> i will update this issue when there is progress

**Comment by letoams** [comment]
On Mon, 1 Jul 2019, jeffguorg wrote:

> Sure. We checked the product brief and datasheets. Intel X520 10Gbps NICs should support nic offload and esp encapsulation. and we have a X520(82599ES).
> is there anyway i can check whether it support that in the system? i scammed across dmesg but didn't recognize something useful.
> 
> ref on Intel X520:
> https://www.intel.com/content/www/us/en/embedded/products/networking/82599-10-gbe-controller-datasheet.html?asset=2377
> https://www.kernel.org/doc/html/latest/networking/device_drivers/intel/ixgbe.html

Please check with libreswan 3.28 or 3.29, as those versions better print
the support detected from either nic or kernel.

You should see a message:

Kernel supports NIC esp-hw-offload

And for the nic you should see a message about esp-hw-offload as well.
It should also be visible in "ipsec status".

with plutodebug=all you should also see a message:

                 dbg("NIC esp-hw-offload offload for connection '%s' enabled on interface %s",

**Comment by jeffguorg** [comment]
weird...

i pulled the source code from master branch and did make install. 
nic-offload is set to no and the bandwidth get to 4Gbps. nearer to our goal
but the bandwidth drop to 27M when i set nic-offload to yes or auto.

and i checked ip xfrm state. after i set nic-offload to auto, there is two line says ```crypt offload: .... dev <interface>...```, one for dir in and one for out.
so i'm sure the nic is recognized and utilized but it's not working properly

and i did some research and found the NIC support only aes_gcm128-null for esp encapsulation and that is exactly what we use in the config.

**Comment by letoams** [comment]
Try setting replay-window=0

Paul

**Comment by jeffguorg** [comment]
just minor improvement: 27.1M -> 28.4M

/proc/net/xfrm_state is all zero.

**Comment by letoams** [comment]
On Tue, 2 Jul 2019, jeffguorg wrote:

> just minor improvement: 27.1M -> 28.4M

Odd, there must be other issues. do you have proper max_cpu_qlen set?
See dmesg.  It should be at least 1000.

does "ip xfrm state" show the offload is part of the IPsec SA ?

**Comment by jeffguorg** [comment]
hi, we tested some configuration last day but did not make any progress. 

just confirmed that the issue occured when nic-offload is enabled on the sending side and collected some logs and output.

> Odd, there must be other issues. do you have proper max_cpu_qlen set?
See dmesg.  It should be at least 1000.

sure. ```[    5.628444] cryptd: max_cpu_qlen set to 4096```. it's the first thing i did when i tried to improve the bandwidth

> does "ip xfrm state" show the offload is part of the IPsec SA ?

yes. it looks like this:

```
src 10.3.0.2 dst 10.3.0.1
	proto esp spi 0x262e4754 reqid 16389 mode tunnel
	replay-window 32 flag af-unspec
	aead rfc4106(gcm(aes)) 0xed2897c7401393ea4fb79d72b838729ddbc22bd4 128
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	crypto offload parameters: dev ens1f0 dir in
src 10.3.0.1 dst 10.3.0.2
	proto esp spi 0x777af742 reqid 16389 mode tunnel
	replay-window 32 flag af-unspec
	aead rfc4106(gcm(aes)) 0xe84b019facd525d9cf0b91ae7da8f3968db877f3 128
	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
	crypto offload parameters: dev ens1f0 dir out

```
another weird thing i found is that in pluto.log, it says: 
```
...
9:48:26.832458: adding interface ens1f0/ens1f0 (esp-hw-offload not supported by kernel) 10.3.0.1:500
Jul  3 09:48:26.832483: | NAT-Traversal: Trying sockopt style NAT-T
Jul  3 09:48:26.832488: | NAT-Traversal: ESPINUDP(2) setup succeeded for sockopt style NAT-T family IPv4
Jul  3 09:48:26.832492: adding interface ens1f0/ens1f0 10.3.0.1:4500
...
```
but actually ethtool recognizes esp-hw-offload:
```
...
esp-hw-offload: on
esp-tx-csum-hw-offload: on
...
```
one of my colleagues guessed it's related to igb driver. is that possible? 


and here is pluto debugging log and other system info you may need for diagnosis, tell me if you need other information:

- [dmesg.log](https://github.com/libreswan/libreswan/files/3358843/dmesg.log)
- [ipsec-status.log](https://github.com/libreswan/libreswan/files/3358845/ipsec-status.log)
- [pluto-offload-off.log](https://github.com/libreswan/libreswan/files/3358847/pluto-offload-off.log)
- [pluto-offload-on.log](https://github.com/libreswan/libreswan/files/3358848/pluto-offload-on.log)
- [ethtool-ens1f0.log](https://github.com/libreswan/libreswan/files/3358885/ethtool-ens1f0.log)

**Comment by letoams** [comment]
On Thu, 4 Jul 2019, jeffguorg wrote:

>       does "ip xfrm state" show the offload is part of the IPsec SA ?
> 
> yes. it looks like this:
> 
> src 10.3.0.2 dst 10.3.0.1
> 	proto esp spi 0x262e4754 reqid 16389 mode tunnel
> 	replay-window 32 flag af-unspec
> 	aead rfc4106(gcm(aes)) 0xed2897c7401393ea4fb79d72b838729ddbc22bd4 128
> 	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
> 	crypto offload parameters: dev ens1f0 dir in
> src 10.3.0.1 dst 10.3.0.2
> 	proto esp spi 0x777af742 reqid 16389 mode tunnel
> 	replay-window 32 flag af-unspec
> 	aead rfc4106(gcm(aes)) 0xe84b019facd525d9cf0b91ae7da8f3968db877f3 128
> 	anti-replay context: seq 0x0, oseq 0x0, bitmap 0x00000000
> 	crypto offload parameters: dev ens1f0 dir out

Okay, so that looks to be working. So it seems this is either an OS
tuning matter, or a driver issue.

> another weird thing i found is that in pluto.log, it says:
> ...
> 9:48:26.832458: adding interface ens1f0/ens1f0 (esp-hw-offload not supported by kernel) 10.3.0.1:500

That was a bug in our printing, you can ignore that.

Paul

**Comment by letoams** [comment]
I still see the replay window being 32. I'm not sure if that is a display error of the kernel or not. If you set it to 0, it should be disabled. Alternatively, perhaps set it to 1024 ?

If you get out of order packets and only have a window of 32, things could go very slow indeed. Although then if you test with iperf or something, it should show you all the packet drop.

**Comment by jeffguorg** [comment]
Hi

The two machine(actually it's [supermicro twins server](https://www.supermicro.com/ja/products/system/1U/1028/SYS-1028TP-DTR.cfm), two server in a single chassis) is now directly connected by optical fibers, without any intermediate devices like switches or routers. it's a testing environment in our office. so there should not be out of order packets, at least not much. 

still i set the replay-window to 1024, but it did not change anything

**Comment by antonyantony** [comment]
you are using "mode tunnel"? I noticed in your outputs. 
I wonder if the ixgbe offload support tunnel mode, afik it only support transport mode. However, what is confusing is you don't get any error, I think automagic is lying there.

I recollect the Shannon mentioning these cards only support transport mode. Note transport in the blog you mentioned, also "ref on Intel X520:" mentions only transport mode. He mentioned the card may support tunnel mode receiving,  never mentioned sending.

Here is Netdev talk 2017. He also gave an update at linux IPsec workshop the following year. No tunnel mode there also.

https://youtu.be/AOEpCRsBTIo?t=1507  (~25:00) I guess better error reporting would be nice.
 
As of 2017/2018 tunnel mode was a "Todo" iteam, with low probability of finishing.

https://netdevconf.org/2.2/slides/klassert_ipsec_workshop02.pdf  #17 

PS: No operational experience here, just slideware and listing to talks.

**Comment by jeffguorg** [comment]
@antonyantony 

hey man, THANKS A LOT, you save my ass. this problem looks like voodoo to me. 

when i set type=transport, and checked ip xfrm state has transport mode, the bandwidth directly runs like 9.1Gbps or a little higher. 

i'd better figure out what's the difference between them. 

@antonyantony @letoams 
Thanks you all! 
wish i could join a team like you someday and change the world


------------------
edit:

and yes.... it's weird that both of libreswan and kernel didn't report any errors. i will read the kernel docs and see if i can find anything

**Comment by jeffguorg** [comment]
the weird issue just came back after i setup a GRE tunnel.

it runs 5.x Gbps without nic-offload and 9.x Gbps with nic-offload. this part of setup is perfect. 

but transport mode doesn't support routing. right? i need to route packets from two subnets to each other, so i setup a GRE tunnel over ipsec, with NetworkManager. then the bandwidth on the tunnel dropped to 20~30 Mbps. again. :sob: 

config looks like this:

```
sudo tee /etc/NetworkManager/system-connections/gre <<EOF
[connection]
id=gre
uuid=ae8f448b-8e77-4d0c-b2fb-e3a562ed8aaa
type=ip-tunnel
interface-name=gre
permissions=
timestamp=1562913160

[ip-tunnel]
local=${LOCAL}
mode=2
remote=${REMOTE}

[ipv4]
address1=${ADDR}
dns-search=
ignore-auto-dns=true
ignore-auto-routes=true
method=manual

[ipv6]
addr-gen-mode=stable-privacy
dns-search=
method=ignore
EOF
```

LOCAL and REMOTE is IPSec gateways. 

mtu on GRE is automatically set to 1280, when it's around 1480 or 1500 on IPSec. just looks normal to me. 

top doesn't show any busy process or thread.

**Comment by letoams** [comment]
On Tue, 16 Jul 2019, jeffguorg wrote:

> it runs 5.x Gbps without nic-offload and 9.x Gbps with nic-offload. this part of setup is perfect.
> 
> but transport mode doesn't support routing. right? i need to route packets from two subnets to each other, so i setup a GRE tunnel over ipsec, with
> NetworkManager. then the bandwidth on the tunnel dropped to 20~30 Mbps. again. ?

Thats right, and people tend to use GRE on top to handle it. Did you
slightly lower your MTU to take account for the smaller packet size
due to the GRE layer?

Paul
